4단계 - 로또(2등)

lsm7179 opened the pull request:
안녕하세요.
강현구님 4단계 로또(2등) 리뷰 요청드립니다.
말씀해주신 피드백 반영했습니다.
꼼꼼하게 봐주셔서 감사합니다 :) 
더 열심히 해볼게요 :fire::fire::fire:

lsm7179 committed:
feat: 당첨번호와 보너스 볼을 가지는 객체 생성

lsm7179 committed:
refacor: 랭크 순위 결과값 메소드 리팩토링

lsm7179 committed:
feat: 당첨번호와 로또를 비교하여 몇개 맞췄는지 랭크값 확인하는 기능

lsm7179 committed:
refator : list를 set으로 변환하는 기능 리팩토링

lsm7179 committed:
feat: 당첨번호와 보너스번호를 가지고 당첨로또를 생성하는 기능

lsm7179 committed:
refactor: 피드백 반영

특별한 이유 없는 static메소드 제거
배열 대신 리스트로 변경
뷰에 의존적인 toString메소드 제거
테스트메소드명 적절하게 변경
ResultTest.java 파일 생성

lsm7179 committed:
feat: 로또 출력 기능 ResultView로 변경

lsm7179 committed:
style: 코드 포맷팅

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -4,6 +4,7 @@
 import java.util.List;
+import java.util.Objects;
 
 public class Lotto {
```
아래 코드를 허용하지 않도록 개선해보아요

```java
final Lotto lotto = new Lotto(Arrays.asList(1, 1, 1, 1, 1, 1));
```

kang-hyungu commented on src/main/java/lotto/factory/LottoCreateFactory.java:
```diff
@@ -51,4 +51,7 @@ private static boolean validDuplicate(List<Integer> numbers) {
         return set.size() != Lotto.SIZE;
```
팩터리가 유효성 체크 책임도 맡는게 적절할까요?

kang-hyungu commented on src/main/java/lotto/LottoGame.java:
```diff
@@ -34,19 +31,21 @@ private void buyLotto() {
+        ResultView.printLottoList(lottos);
     }
 
     private void setWinningLotto() {
```
setter가 아닌데 메서드명을 set으로 시작할 필요가 없어보이네요
적절한 메서드명 없을까요?

kang-hyungu commented on src/main/java/lotto/view/InputHandler.java:
```diff
@@ -37,7 +36,7 @@ public static int price(String priceText) {
     }
 
-    private static int checkLottoNumber(String textNumber) {
+    public static int checkLottoNumber(String textNumber) {
```
중복 코드가 생겼네요
이미 모델 객체에서 로또 번호를 체크하고 있는데 입력 받을데 게임 규칙을 또 체크할 필요가 있을까요?
입력이 제대로 됐는지만 판단하고 게임 규칙과 관련된 유효성 검사는 모델 객체에서 처리하는게 어떨까요

kang-hyungu commented on src/main/java/lotto/model/Result.java:
```diff
@@ -12,10 +11,10 @@
 
-    public Result(Lottos lottos, Lotto winLotto) {
+    public Result(Lottos lottos, WinningLotto winLotto) {
         this.matchResult = createResultMap();
```
```suggestion
        this.matchResult = Rank.createWinningRanks();
```

이렇게 바꿔도 되지 않나요?
Rank 값을 꺼내서 각각 0으로 초기화 해줄게 아니라 Rank 내부에서 필요한 값 셋팅해서 필요한 값으로 반환해주면 재사용도 가능하지 않을까요?

kang-hyungu replied:
물론 createWinningRanks 메서드는 Map을 반환하도록 수정하셔야합니다.

kang-hyungu commented on src/main/java/lotto/model/WinningLotto.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Objects;
+
+public class WinningLotto {
+    private Lotto winLotto;
```
```suggestion
    private final Lotto winLotto;
```

kang-hyungu commented on src/main/java/lotto/model/WinningLotto.java:
```diff
@@ -0,0 +1,41 @@
+    private LottoNumber bonusNumber;
+
+    public WinningLotto(Lotto winLotto, LottoNumber bonusNumber) {
+        this.winLotto = winLotto;
```
winLotto는 null 허용하나요?

kang-hyungu commented on src/main/java/lotto/factory/LottoCreateFactory.java:
```diff
@@ -11,8 +12,7 @@
 
     public static Lotto createRandomLotto() {
```
팩터리 내부와 테스트 코드에서만 사용하네요
접근제어자를 public으로 열어둘 필요가 있을까요?

kang-hyungu commented on src/test/java/lotto/model/LottosTest.java:
```diff
@@ -7,40 +7,19 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class LottosTest {
```
아래 코드도 허용하나요?
테스트 코드로 검증해보아요

```java
new Lottos(null);
```

kang-hyungu commented on src/test/java/lotto/model/LottoTest.java:
```diff
@@ -24,13 +24,17 @@ void lottoCreate() {
+        winLotto = new WinningLotto(new Lotto(Arrays.asList(1, 2, 3, 4, 5, 6)), new LottoNumber(8));
+        assertThat(winLotto.matchRank(lotto)).isEqualTo(Rank.THIRD);
     }
 
```
matchNumber 메서드는 의도대로 동작할까요?
빠진 테스트는 없는지 점검해보아요

kang-hyungu replied:
Price도 점검해보는게 좋겠네요

```java
Price.numberPurchases(-100);
Price.numberPurchases(100);
Price.numberPurchases(0);
```

kang-hyungu commented on src/test/java/lotto/model/ResultTest.java:
```diff
@@ -0,0 +1,36 @@
+        BigInteger purchase = Price.totalPurchase(lottos.size());
+
+        assertThat(lottos.size()).isEqualTo(2);
+        assertThat(purchase.toString()).isEqualTo("2000");
```
Price의 기능을 왜 ResultTest에서 검증하나요?
PriceTest에서 테스트 해야되지 않을까요?
위에 로또 사이즈도 마찬가지입니다.

kang-hyungu reviewed (changes requested):
안녕하세요 승민님
몇 가지 피드백 남겼습니다. 테스트 코드도 같이 점검해보면 좋겠네요
궁금한 점 생기면 dm 주시거나 코멘트 남겨주세요

lsm7179 committed:
refactor: 리뷰사항 반영

로또번호가 중복되지 않도록 처리
모델 객체에서 게임 규칙과 관련된 유효성 검사 책임맡도록 변경
메소드 역할에 맞는 적절한 메서드명으로 변경
중복 코드 제거

lsm7179 committed:
refactor: 리뷰사항 반영

테스트 코드로 검증
Price 테스트 코드 추가
null 값 error 처리

lsm7179 committed:
test: LottoNumberTest 테스트 코드 추가

kang-hyungu commented on src/main/java/lotto/LottoGame.java:
```diff
@@ -34,19 +31,21 @@ private void buyLotto() {
     }
 
-    private void setWinningLotto() {
+    private void enterWinningLotto() {
```
값을 입력 받는 것이니 inputWinningLotto를 제안해봅니다.

kang-hyungu commented on src/main/java/lotto/factory/LottoCreateFactory.java:
```diff
@@ -32,23 +32,18 @@ public static Lotto createRandomLotto() {
-    private static boolean validDuplicate(List<Integer> numbers) {
-        Set<Integer> set = new HashSet<>(numbers);
-        return set.size() != Lotto.SIZE;
+    private static Lotto createLotto(List<Integer> numbers) {
```
이 메서드도 불필요해보이네요.
한 번 더 메서드로 감쌀 필요가 있을까요?

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -18,34 +17,61 @@ public Lotto(List<Integer> numbers) {
         int result = 0;
         for (LottoNumber lottoNumber : lottoNumbers) {
-            result += winLotto.compare(lottoNumber) ? 1 : 0;
+            result += lotto.compare(lottoNumber) ? 1 : 0;
```
lotto.compare(lottoNumber) ? 1 : 0;
이 부분도 메서드로 추출하면 어떨까요?
왜 로또번호를 비교해서 1이나 0을 반환하는지 이해하는데 도움이 되지 않을까요?

lsm7179 replied:
이 부분 밑에처럼 바꿔봤는데 이 방법이 1이나 0을 반환하는지 도움이 될까요?
메소드명을 뭐라고 짓는게 좋을지 어렵네요..
저는 이전코드가 for문안에 있어서 이해하기 더 편한거 같은데 어떻게 생각하시나요?
````
 public static final int INCREMENT_BY_ONE = 1;
 public static final int INCREMENT_BY_ZERO = 1;

 public int matchNumber(Lotto lotto) {
        int result = 0;
        for (LottoNumber lottoNumber : lottoNumbers) {
            result += increment(lotto, lottoNumber);
        }
        return result;
    }

 private int increment(Lotto lotto, LottoNumber lottoNumber) {
        if(lotto.compare(lottoNumber)){
            return INCREMENT_BY_ONE;
        }
        return INCREMENT_BY_ZERO;
    }
````

아니면 이렇게 한줄로 메소드를 추출도 해보았습니다.
어떤게 더 좋은 방법일까요?..
```` 
private int increment(Lotto lotto, LottoNumber lottoNumber) {
         return lotto.compare(lottoNumber) ? INCREMENT_BY_ONE : INCREMENT_BY_ZERO;
    }
````

kang-hyungu replied:
코드를 처음 보는 개발자가 increment 라는 메서드명을 읽었을때 로또, 로또 번호를 넘겨주면 무엇이 증가한다는 것이지? 라고 생각하지 않을까요?
if절로 할지 삼항연상자를 할지는 팀에서 정한 방침을 따르면 될 것 같네요.

lsm7179 replied:
> 무엇이 증가한다는

지금 드는 생각은 
로또 맞춘갯수가 증가 matchNumberIncrement
또는 당첨번호 밎춘 갯수 증가 winningCountIncreament
이렇게 인데 matchNumberIncrement가 '로또 맞춘갯수 증가' 라는 말이 더 적절한거 같아서
matchNumberIncrement 좋아 보입니다.

lsm7179 replied:
> 코드를 처음 보는 개발자가 increment 라는 메서드명을 읽었을때 로또, 로또 번호를 넘겨주면 무엇이 증가한다는 것이지? 라고 생각하지 않을까요? if절로 할지 삼항연상자를 할지는 팀에서 정한 방침을 따르면 될 것 같네요.

지금 드는 생각은
로또 맞춘갯수가 증가 matchNumberIncrement
또는 당첨번호 밎춘 갯수 증가 winningCountIncreament
이렇게 인데 '로또 맞춘갯수 증가' 라는 말이 더 적절한거 같아서
matchNumberIncrement 좋아 보입니다.

kang-hyungu commented on src/main/java/lotto/model/Rank.java:
```diff
@@ -29,16 +28,35 @@ public int getWinningMoney() {
-    public static List<Rank> createWinningRanks() {
-        return Stream.of(FIRST, SECOND, THIRD, FOURTH)
-                .sorted(Comparator.comparing(Rank::getCountOfMatch))
+    public static Map<Rank, Integer> createRankMap() {
```
반환 타입에 map이 있으니 메서드명에 또 써줄 필요는 없어보입니다.

kang-hyungu reviewed (approved):
승민님 피드백 반영 잘 해주셨어요
추가 코멘트는 다음 미션에서 적용해주세요
pr은 머지 처리하겠습니다. 수고하셨습니다.
